### PR TITLE
SSH container support

### PR DESCRIPTION
**Pull request by mfalesni**
DONE:
* Prefixing with docker exec with proper quoting
* file upload and download via intermediate location
* Pass the fact it is running in container using env.yaml

### CODE REVIEW

**Comment by psav** [comment]
Could we just have the container vm have a "service" running on a port - if we detect that port, then we know we are hitting a containerized version?

**Comment by dajoRH** [comment]
Lint report for commit 0dbf6939eeaf79f0e51cc0c80f7333638b282d2f:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by mfalesni** [comment]
@psav Possibly, it could be something REST that even serves you username and password :D

**Comment by anewmanRH** [comment]
A file on disk I think would work just as well.

**Comment by psav** [comment]
@anewmanRH the problem here is we have nobbled the SSH client to be container or no.....so we need to know that before we can use it :)

**Comment by mfalesni** [comment]
Exactly.

**Comment by psav** [comment]
A banner check on SSH could be sufficient

**Comment by mfalesni** [comment]
Yeah, I have to figure that out, that sounds like a good idea.

**Comment by dajoRH** [comment]
Lint report for commit 81ecfb3f9c6f1b5ae335b7da2643849c456af0de:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by mfalesni** [comment]
@RonnyPfannschmidt @psav This will probably have to be aligned somehow with the appliance stack efforts. This needs only one extra parameter by the way. I guess the pytest_store could have its own configure which would configure the appliance IP and ports and it would come first in the order of plugins.

**Comment by dajoRH** [comment]
Lint report for commit ae0bb8fd56583e67d14dc67528791c416f14ec82:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8864ede5c1eb5763c516426a8548c0a274d87cde:

Everything seems all right :smile: .
*CFME QE Bot*

File: utils/ssh.py [hunk 1]
```diff
@@ -135,10 +148,14 @@ def get_transport(self, *args, **kwargs):
             self.connect()
         return super(SSHClient, self).get_transport(*args, **kwargs)
 
-    def run_command(self, command, timeout=RUNCMD_TIMEOUT, reraise=False):
+    def run_command(self, command, timeout=RUNCMD_TIMEOUT, reraise=False, ensure_host=False):
         if isinstance(command, dict):
             command = version.pick(command)
         logger.info("Running command `%s`", command)
+        if self._container and not ensure_host:
```
Comment (psav) [inline utils/ssh.py]: ```self.is_container``` here for readability???

Comment (mfalesni) [inline utils/ssh.py]: It somehow makes sense to me to use the same variable for check and for the value later ... dunno. Of course I can change it.

File: utils/ssh.py [hunk 2]
```diff
@@ -208,13 +225,28 @@ def run_rake_command(self, command, timeout=RUNCMD_TIMEOUT):
 
     def put_file(self, local_file, remote_file='.', **kwargs):
         logger.info("Transferring local file %s to remote %s", local_file, remote_file)
-        return SCPClient(self.get_transport(), progress=self._progress_callback).put(
-            local_file, remote_file, **kwargs)
+        if self.is_container:
+            tempfilename = '/share/temp_{}'.format(fauxfactory.gen_alpha())
+            scp = SCPClient(self.get_transport(), progress=self._progress_callback).put(
+                local_file, tempfilename, **kwargs)
+            self.run_command('mv {} {}'.format(tempfilename, remote_file))
```
Comment (psav) [inline utils/ssh.py]: How does this "mv" it into the container? Do you have to know the remote path to the container? ie a mounted volume?

Comment (mfalesni) [inline utils/ssh.py]: It assumes there is /share directory visible both in container and the host VM. A prerequisite.

Comment (psav) [inline utils/ssh.py]: OK, so you dump it in share, and then retriving it is using the same name, but is run inside the docker context...got it

Comment (mfalesni) [inline utils/ssh.py]: Exactly. Basically one additional step with a random name.

The docs I pushed clarify the prerequisities for the VM that hosts the docker container.

**Comment by dajoRH** [comment]
Lint report for commit a380b651ee6fdc9c075fc07614d36d0860f3bc9c:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by psav** [comment]
:cake::punch:

**Pull request merged**